src/fossa_cli.rs

```diff
@@ -48,29 +57,106 @@ pub enum Error {
+        Some(PathBuf::from(command))
+    } else {
+        None
+    };
+
+    let desired_version = latest_release_version().await?;
+    match current_path {
+        Some(current_path) => download_if_old(ctx, current_path, desired_version).await,
+        None => download(ctx, desired_version).await,
     }
+}
 
-    // if it does not exist in the data root, then check to see if it is on the path.
-    if check_command_existence(&PathBuf::from(&command)).await {
-        return Ok(PathBuf::from(command));
-    };
+/// If the cli exists locally, then compare the version of the local CLI to the latest release,
+/// and download it if it is different.
+/// If there are any errors while finding the local version, then just download it.
+#[tracing::instrument]
+async fn download_if_old(
+    ctx: &AppContext,
+    current_path: PathBuf,
+    desired_version: String,
+) -> Result<PathBuf, Error> {
+    if let Ok(local_version) = local_version(&current_path).await {
+        if local_version == desired_version {
+            debug!(
+                "local version of fossa-cli at {} matches desired version of {}",
+                current_path.display(),
+                desired_version
+            );
+            Ok(current_path)
+        } else {
+            debug!(
+                "local version of fossa-cli at {} has version of {}, which does not match desired version of {}. Downloading new version.",
+                current_path.display(),
+                local_version,
+                desired_version,
+            );
+            download(ctx, desired_version).await
+        }
+    } else {
+        debug!(
+            "Error while getting version from local fossa-cli at {}. Downloading new version",
+            current_path.display()
+        );
+        download(ctx, desired_version).await
+    }
+}
 
-    // if it is not in either location, then download it
-    download(ctx.data_root())
-        .await
-        .describe("fossa-cli not found in your path, attempting to download it")
+#[tracing::instrument(fields(fossa_version_output))]
+async fn local_version(current_path: &PathBuf) -> Result<String, Error> {
+    let output = Command::new(current_path)
+        .arg("--version")
+        .output()
+        .context(Error::RunLocalFossaVersion)?;
+    if !output.status.success() {
+        return Error::RunLocalFossaVersion.wrap_err().into_report();
+    }
+
+    // the output will look something like "fossa-cli version 3.7.2 (revision 49a37c0147dc compiled with ghc-9.0)"
+    let output = String::from_utf8(output.stdout).context(Error::RunLocalFossaVersion)?;
+    span_record!(fossa_version_output, debug output);
+
+    let version = output
+        .strip_prefix("fossa-cli version ")
+        .ok_or(Error::ParseLocalFossaVersion(output.clone()))?;
+    let version = version
+        .split(' ')
+        .next()
+        .ok_or(Error::ParseLocalFossaVersion(output.clone()))?;
```

Comment:
[nit] Any time we must potentially run a function in the error case, prefer the lazy variant:

```suggestion
    let version = output
        .strip_prefix("fossa-cli version ")
        .ok_or_else(|| Error::ParseLocalFossaVersion(output.clone()))?;
    let version = version
        .split(' ')
        .next()
        .ok_or_else(|| Error::ParseLocalFossaVersion(output.clone()))?;
```